Sources/SwiftSyntax/SyntaxParser.swift

```diff
@@ -0,0 +1,225 @@
+typealias CTokenKind = swiftparse_token_kind_t
+typealias CParseLookupResult = swiftparse_lookup_result_t
+
+/// A list of possible errors that could be encountered while parsing a
+/// Syntax tree.
+public enum ParserError: Error, CustomStringConvertible {
+  /// Parser created an invalid syntax data. That should never occur under
+  /// normal circumstances, and it should be reported as a bug.
+  case invalidSyntaxData
+
+  public var description: String {
+    switch self {
+    case .invalidSyntaxData:
+      return "parser created invalid syntax data"
+    }
+  }
+}
+
+/// Provides a mechanism for the parser to skip regions of an incrementally
+/// updated source that were already parsed during a previous parse invocation.
+public protocol IncrementalParseLookup {
+  /// Does a lookup to see if the current source `offset` should be associated
+  /// with a known `Syntax` node and its region skipped during parsing.
+  ///
+  /// The implementation is responsible for checking whether an incremental edit
+  /// has invalidated the previous `Syntax` node.
+  ///
+  /// - Parameters:
+  ///   - offset: The byte offset of the source string that is currently parsed.
+  ///   - kind: The `SyntaxKind` that the parser expects at this position.
+  /// - Returns: A `Syntax` node from the previous parse invocation,
+  ///            representing the contents of this region, if it is still valid
+  ///            to re-use. `nil` otherwise.
+  func lookUp(_ offset: Int, kind: SyntaxKind) -> Syntax?
+}
+
+/// Namespace for functions to parse swift source and retrieve a syntax tree.
+public struct SyntaxParser {
+  /// Parses the string into a full-fidelity Syntax tree.
+  ///
+  /// - Parameters:
+  ///   - source: The source string to parse.
+  ///   - parseLookup: Optional mechanism for incremental re-parsing.
+  /// - Returns: A top-level Syntax node representing the contents of the tree,
+  ///            if the parse was successful.
+  /// - Throws: `ParserError`
+  public static func parse(
+    source: String,
+    parseLookup: IncrementalParseLookup? = nil
+  ) throws -> SourceFileSyntax {
+    // Get a native UTF8 string for efficient indexing with UTF8 byte offsets.
+    // If the string is backed by an NSString then such indexing will become
+    // extremely slow.
+    var utf8Source = source
+    if !utf8Source.isNativeUTF8 {
+      utf8Source.makeNativeUTF8()
+    }
+
+    let rawSyntax = parseRaw(utf8Source, parseLookup: parseLookup)
+
+    guard let file = makeSyntax(rawSyntax) as? SourceFileSyntax else {
+      throw ParserError.invalidSyntaxData
+    }
+    return file
+  }
+
+  /// Parses the file `URL` into a full-fidelity Syntax tree.
+  ///
+  /// - Parameters:
+  ///   - url: The file URL to parse.
+  /// - Returns: A top-level Syntax node representing the contents of the tree,
+  ///            if the parse was successful.
+  /// - Throws: `ParserError`
+  public static func parse(url: URL) throws -> SourceFileSyntax {
```

Comment:
Nitpick: I prefer this as `parse(_ url: URL)`, otherwise you're just restating type information.

Reply:
Now there's `parse(url: URL)` and `parse(source: String)`. My slight concern here is that if labels are removed then if you see code like `parse(source)` it's not obviously clear if it is doing a file URL parse or a string one.

Reply:
IMO there would be `parse(_:)` that takes a URL, and `parse(source:)` that takes source code. That would remove the ambiguity and give semantic meaning to the String parameter (to distinguish it from a path), while also not repeating type information in the `URL` case.

Reply:
Make sense, done.

Reply:
Thanks!